### PR TITLE
AMQP-675: RabbitTemplate - Add Typed Returns

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/AMQP-675

Add variants to `RabbitTemplate` `receiveAndConvert` and `convertSendAndReceive` methods
that take a `ParameterizedTypeReference` argument.

Also `AsyncRabbitTemplate`.

Add `SmartMessageConverter` (similar to other spring projects) that take a conversion hint.
Implement this interface in the `Jackson2JsonMessageConverter`.

Also, pull up the method variants that take `CorrelationData` to `RabbitOperations`.

### CODE REVIEW

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/ComplexTypeJsonIntegrationTests.java [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+		Foo<Bar<Baz, Qux>> foo = new Foo<>();
+		Bar<Baz, Qux> bar = new Bar<>();
+		bar.setaField(new Baz("foo"));
+		bar.setbField(new Qux(42));
+		foo.setField(bar);
+		return foo;
+	}
+
+	/*
+	 * Covers all flovors of convertSendAndReceiveAsType
```
Comment (artembilan) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/ComplexTypeJsonIntegrationTests.java]: Typo: fl**a**vors

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/ComplexTypeJsonIntegrationTests.java [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.test.context.junit4.SpringRunner;
+
+/**
+ * @author Gary Russell
+ * @since 2.0
+ *
+ */
+@RunWith(SpringRunner.class)
+public class ComplexTypeJsonIntegrationTests {
```
Comment (artembilan) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/annotation/ComplexTypeJsonIntegrationTests.java]: `@DirtiesContext` to stop listener container in the end ?

File: src/reference/asciidoc/amqp.adoc [hunk 3]
```diff
@@ -2537,6 +2540,20 @@ For this reason, the infrastructure provides the `targetObject` message property
 converter to determine the type.
 ====
 
+[[json-complex]]
+====== Converting From a Message With RabbitTemplate
+
+As mentioned above, type information is conveyed in message headers to assist the converter when converting from a message.
+This works fine in most cases, but when using generic types, it can only convert simple objects and known "container" objects (lists, arrays, maps).
+Starting with __version 2.0__, the `Jackson2JsonMessageConverter` implements `SmartMessageConverter` which allows it to be used with the new `Rabbit` template methods that take a `ParameterizedTypeReference` argument; this allows conversion of complex generic types.
```
Comment (artembilan) [inline src/reference/asciidoc/amqp.adoc]: Maybe "`RabbitTemplate` methods " would look better ?

File: src/reference/asciidoc/amqp.adoc [hunk 4]
```diff
@@ -2695,11 +2712,15 @@ Those methods are quite useful for request/reply scenarios since they handle the
 
 Similar request/reply methods are also available where the `MessageConverter` is applied to both the request and reply.
 Those methods are named `convertSendAndReceive`.
-See the Javadoc of `AmqpTemplate` for more detail.
+See the the Javadoc of `AmqpTemplate` for more detail.
```
Comment (artembilan) [inline src/reference/asciidoc/amqp.adoc]: Double `the` 😄

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java [hunk 5]
```diff
@@ -399,6 +403,57 @@ public RabbitMessageFuture sendAndReceive(String exchange, String routingKey, Me
...
 	public <C> RabbitConverterFuture<C> convertSendAndReceive(String exchange, String routingKey, Object object,
 			MessagePostProcessor messagePostProcessor) {
+		return convertSendAndReceive(exchange, routingKey, object, messagePostProcessor, true);
+	}
+
+	@Override
+	public <C> RabbitConverterFuture<C> convertSendAndReceiveAsType(Object object,
+			ParameterizedTypeReference<C> responseType) {
+		return convertSendAndReceiveAsType(this.template.getExchange(), this.template.getRoutingKey(), object,
+				(MessagePostProcessor) null, responseType);
```
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: Well - perhaps when they fix "unused imports" and "unused variables" like that one you left in yesterday 😄

Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/AsyncRabbitTemplate.java]: He-he! There are both of them, but they need human being interaction and attention. 
As well as this "unnecessary cast" 😉 

OK. Never mind.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-amqp/commit/b412e6f3cebd09080beb5762ef972803b41ca785

**Pull request closed without merging**